.dependency-cruiser.js

```diff
@@ -183,7 +238,140 @@ module.exports = {
           'npm-peer'
         ]
       }
-    }
+    },
+
+    /* application-specific architecture rules */
+    {
+      name: 'dependencies-of-shared',
+      severity: 'error',
+      from: {
+        path: dir('src/shared')
+      },
+      to: {
+        pathNot: [
+          dir('src/shared'),
+          dir('src/fs'),
+          ...npmDeps(
+            'node-fetch',
+            'markdown-escape',
+            'bn.js',
+          )
+        ]
+      }
+    },
+    {
+      name: 'dependencies-of-fs',
+      severity: 'error',
+      from: {
+        path: dir('src/fs')
+      },
+      to: {
+        pathNot: [
+          // TODO(SRe): is this okay?
+          dir('src/shared'), // 'shared' and 'fs' form a cycle
+          dir('src/fs'),
+          ...modules(
+            'path',
+            'fs',
+            'util',
+          ),
+        ]
+      }
+    },
+    {
+      name: 'dependencies-of-client',
+      severity: 'error',
+      from: {
+        path: dir('src/client'),
+        pathNot: [
+          file('src/client/webpack.config.js'),
+          testFilesWithinDir('src/client'),
+        ],
+      },
+      to: {
+        pathNot: [
+          dir('src/client'),
+          dir('src/shared'),
+          file('data/libraries/libraries.json'),
+          ...npmDeps(
```

Comment:
Maybe we should just allow all npm packages here if possible. But I'm undecided.

Reply:
In my opinion, we should not allow the client to depend on packages intended to be used with node (like e.g. `node-fetch`). But maybe the bureaucracy of listing everything the client is allowed to depend on is not worth the effort.

Reply:
Good point. One more reason to try the "workspaces" approach I mentioned below. Then the client will have its own `package.json` with non-node modules only.